src/lib/krb5/asn.1/asn1_encode.c

```diff
@@ -191,9 +193,12 @@ k5_asn1_decode_int(const uint8_t *asn1, size_t len, intmax_t *val)
     if (len == 0)
         return ASN1_BAD_LENGTH;
     n = (asn1[0] & 0x80) ? -1 : 0;
-    /* Check length; allow extra octet if first octet is 0. */
-    if (len > sizeof(intmax_t) + (asn1[0] == 0))
+    /* Check length. */
+    if (len > sizeof(intmax_t))
```

Comment:
On further review, this function may need more attention to avoid undefined behavior.

C99 6.5.7 defines left shift for signed integers as follows: "If E1 has a signed type and nonnegative value, and E1 × 2
E2 is representable in the result type, then that is the resulting value; otherwise, the behavior is undefined."

Eliminating the extra byte allowance prevents an encoding like 00 FF FF FF FF FF FF FF FF from shifting into the sign bit (which is undefined) on the last iteration of the for loop.  However, when the integer is negative, we are also triggering undefined behavior at every step.

Heimdal uses multiply-by-256-and-add rather than shift-left-8-and-bitwise-or.  That might be the simplest fix.

(Apparently C++20 has a more complete specification for left shift.  But of course this isn't C++ code, and C18 appears to have the same specification as C99.)

Reply:
That’s unfortunate :disappointed:.  Another option would be to use an unsigned type instead.